cli/sdo/voucher.go

```diff
@@ -151,6 +151,126 @@ type ImportResponse struct {
+	}
+
+	creds := cliutils.OrgAndCreds(org, userCreds)
+	method := http.MethodGet
+	apiMsg = method + " " + sdoURL
+	httpClient := cliutils.GetHTTPClient(config.HTTPRequestTimeoutS)
+
+	resp := cliutils.InvokeRestApi(httpClient, method, sdoURL, creds, requestBodyBytes, "SDO Owner Service", apiMsg)
+	defer resp.Body.Close()
+	httpCode := resp.StatusCode
+	cliutils.Verbose(msgPrinter.Sprintf("HTTP code: %d", httpCode))
+
+	respBodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("failed to read exchange body response from %s: %v", apiMsg, err))
+	}
+	if httpCode != 200 {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("bad HTTP code %d from %s: %s", httpCode, apiMsg, string(respBodyBytes)))
+	}
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+	}
+
+	return respBodyBytes, apiMsg
+}
+
+// called when the -l flag is used to list the full voucher
+func listFullVoucher (respBodyBytes []byte, apiMsg string) []byte {
+	msgPrinter := i18n.GetMessagePrinter()
+	cliutils.Verbose(msgPrinter.Sprintf("Listing imported SDO vouchers."))
+
+	var output interface{}
+	err := json.Unmarshal(respBodyBytes, &output)
+	if err != nil {
+		cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+	}
+
+	jsonBytes, err := json.MarshalIndent(output, "", cliutils.JSON_INDENT)
+	if err != nil {
+		cliutils.Fatal(cliutils.JSON_PARSING_ERROR, msgPrinter.Sprintf("failed to marshal 'hzn exchange service list' output: %v", err))
+	}
+
+	return jsonBytes
+}
+
+// list the all the uploaded SDO vouchers, or a single voucher
+func VoucherList(org, userCreds, voucher string, namesOnly bool) {
+	msgPrinter := i18n.GetMessagePrinter()
+	cliutils.Verbose(msgPrinter.Sprintf("Listing imported SDO vouchers."))
+
+	// call the ocs-api to get the uploaded vouchers
+	var respBodyBytes []byte
+	var apiMsg string
+	respBodyBytes, apiMsg = getVouchers(org, userCreds, apiMsg, voucher)
+
+	// no voucher was specified, list all uploaded vouchers
+	if voucher == "" {
+		output := []string{}
+		err := json.Unmarshal(respBodyBytes, &output)
+		if err != nil {
+			cliutils.Fatal(cliutils.HTTP_ERROR, msgPrinter.Sprintf("json unmarshalling HTTP response '%s' from %s: %v", string(respBodyBytes), apiMsg, err))
+		}
+
+		jsonBytes, err := json.MarshalIndent(output, "", cliutils.JSON_INDENT)
+		if err != nil {
+			cliutils.Fatal(cliutils.JSON_PARSING_ERROR, msgPrinter.Sprintf("failed to marshal 'hzn exchange service list' output: %v", err))
+		}
+
+		// list only the uuid's of imported vouchers
+		if namesOnly {
+			fmt.Printf("%s\n", jsonBytes)
+
+		} else { // list full details of all imported vouchers
+			for i := range output {
+				respBodyBytes, apiMsg = getVouchers(org, userCreds, apiMsg, output[i])
```

Comment:
This will be inefficient for a lot of vouchers. You should add a url parameter to the api for getting full voucher content: `GET /api/vouchers?long=true`

This can be done in a separate PR

Reply:
I wanted to get this in a little earlier because I knew something was going to be brought up about this, and that sounds like a much better solution